skiko/src/iosMain/kotlin/org/jetbrains/skiko/redrawer/MetalRedrawer.ios.kt

```diff
@@ -33,12 +35,62 @@ internal class MetalRedrawer(
     private val queue = device.newCommandQueue() ?: throw IllegalStateException("Couldn't create Metal command queue")
     private var currentDrawable: CAMetalDrawableProtocol? = null
     private val metalLayer = MetalLayer()
+
+    /**
+     * Indicates whether draw was already dispatched during current vsync interval,
+     * so any extra work should be scheduled on next frame
+     */
+    private var canDrawOnCurrentVsync = false
+
+    /**
+     * UITouch events are dispatched right before next CADisplayLink callback by iOS.
+     * It's too late to encode any work for this frame after this happens.
+     * Any work dispatched before the next CADisplayLink callback should be scheduled after that callback.
+     */
+    fun preventDrawDispatchDuringCurrentVsync() {
+        canDrawOnCurrentVsync = false
+    }
+
+    /**
+     * Needs scheduling displayLink for forcing UITouch events to come at the fastest possible cadence.
+     * Otherwise touch events will come on 60hz rate
+     */
+    var needsProactiveDisplayLink = false
+        set(value) {
+            field = value
+
+            if (value) {
+                caDisplayLink.setPaused(false)
+            }
+        }
+
+    /**
+     * Indicates that there is update that needs redraw, which will be executed once new vsync is coming
+     */
+    private var hasScheduledDrawForNextVsync = false
```

Comment:
Is implicit logic in setter considered a bad practice in Kotlin? In my opinion it prevents impossible states (we can't change amount of touches without causing side effects which are mandatory). Another way would be to it would be to extract it to a separate class and make it explicit and still enforce the invariant within class. But that way we can create a class for every possible invariant existing.

3 bools flags are ok here in my opinion, because there is no impossible combination of them.
Extracting a state machine to some sort of sealed interface would be just more verbose.

Reply:
```
/**
 * Explanation of frame scheduling logic on iOS:
 * `frame` means an interval between two vsyncs
 *
 * needsProactiveDisplayLink
 * |     canDrawOnCurrentVsync
 * |     |     hasScheduledDrawForNextVsync
 * |     |     |
 * false false false  | DispLink is off, there is no draw scheduled, we can't draw anything immediately not to drift timings, because those are not known to us
 * false false true   | DispLink is on, we can't draw right now, but there is scheduled draw for next frame.
 * false true  false  | DispLink is on, but will be off after this frame, we can draw something, if [needRedraw] comes, or just do nothing
 * false true  true   | Impossible state, [needRedraw] would immediately dispatch draw and set both [canDrawOnCurrentVsync] and [hasScheduledDrawForNextVsync] to false
 * true  false false  | DispLink is on, but it's too late in the frame to dispatch draw, no draw is scheduled for next frame
 * true  false true   | DispLink is on, draw was either dispatched or it's too late to dispatch it this frame, draw is scheduled for next frame
 * true  true  false  | DispLink is on, we can draw during this frame immediately, no draw is scheduled on next frame
 * true  true  true   | Impossible state, [needRedraw] would immediately dispatch draw and set both [canDrawOnCurrentVsync] and [hasScheduledDrawForNextVsync] to false
 */
```

Actually I was a bit wrong, some states _are_ impossible, will try to refactor it now.

Reply:
Refactored in latest commit.